doc/hardware/cache/index.rst

```diff
@@ -1,9 +1,166 @@
-.. _cache-guide:
+.. _cache_guide:
 
-Cache Interface
-###############
+Caching Basics
+##############
 
-This is a high-level guide to cache interface and Kconfig options related to
+This section discusses the basics of cache coherency and under what situations a user
+needs to explicitly deal with caching. For more detailed info on Zephyr's caching
+tools, see :ref:`cache_config` for Zephyr Kconfig options or :ref:`cache_api` for
+the API reference. This section primarily focuses on the data cache though there is
+typically also an instruction cache for systems with cache support.
+
+.. note::
+
+  The information here assumes that the architecture-specific MPU support is enabled.
+  See the architecture-specific documentation for details.
+
+When dealing with multiple bus masters that do not share the same cache, the issue
```

Comment:
This paragraph seems likely to scare people. :)  In fact all but one of our SMP devices enforce inter-CPU cache coherence[1] at the level of the architecture ISA.  The one that doesn't, Xtensa, has a somewhat complicated layer that works very hard to present a vision of memory to app code that is as coherent as possible (it makes shared .data/.bss memory uncached, but leaves stacks cached, and has a validation layer to ensure that nothing the kernel recognizes as shared memory gets used via the wrong pointer).  Apps aren't really expected to have to use this API to manage regular data, it's more for use with DMA frameworks, etc...

Basically my vote would be to punt here in the general docs and just state something like "For inter-CPU shared data on SMP systems, Zephyr in general ensures that memory will be seen in a coherent state.  Most applications will only need to use the cache APIs for interaction with external hardware like DMA controllers or foreign CPUs running a different OS image."  And then maybe point to the CONFIG_KERNEL_COHERENCE kconfig docs (or an elaborated RST page) for the special case of Xtensa.

[1] But not memory ordering, which is a related idea but addressed by the barrier API and not the cache layer.

Reply:
That seems reasonable to me. My focus was mainly on DMA but I figured I'd throw SMP concerns in as well (though I haven't researched these in Zephyr very well at all). I'll update the docs to punt on the SMP issues as they're handled somewhat differently, as you said.

Reply:
Added a note about SMP-specific cache coherence and added a link to CONFIG_KERNEL_COHERENCE. I also removed references in the remaining new content to synchronizing shared memory between cores and just focused on DMA and peripherals.